### DISCUSSION TITLE
what is the use of "constraint.sql"?

### DISCUSSION THREAD

**Discussion by hk106609620** [MIT-LCP/mimic-code#1134] [MIMIC-IV]
what is the use of "constraint.sql"?

**Comment by tompollard** [comment]
The [constraint.sql](https://github.com/MIT-LCP/mimic-code/blob/main/mimic-iv/buildmimic/postgres/constraint.sql) script adds constraints to the Postgres database. 

Constraints are rules that must be complied with by the database. For example, the following constraint sets the `subject_id` as a primary key in the `patients` table. A primary key is a special identifier that must be unique. 

Once this constraint is added, your database will raise a "duplicate key" error if you try to insert a value that already appears in the `subject_id` column.

```sql
ALTER TABLE mimic_core.patients
ADD CONSTRAINT patients_pk
  PRIMARY KEY (subject_id);
```

It is not necessary to add these constraints. 

They can be helpful for maintaining database integrity (for example, adding the constraint above would prevent you from accidentally loading duplicate patient data into the `patients` table). 

Constraints may have performance implications, so you may decide that they are unnecessary. MIMIC is essentially a read-only dataset, so integrity is less of a worry than it might be for transactional datasets where there are frequent updates.
